Add new example for section 5

idoleat opened the pull request:
A new directory /examples is added for example codes. .clang-format is copied from sysprog21/lkmpg project.

This draft PR provides a simplified implementation of thread pool. After initializing the thread pool with thread count, jobs can be added. The job queue is a SPMC ring buffer. To keep the implementation minimal, the producer is not protecting, resulting the thread pool can not run automatically when jobs are added. Or the worker may try to get the job before it is fully enqueued.

Padding is added in `thread_pool_t` to avoid false sharing. The number "40" is the sum of size of struct members, including alignment, before the first padding. There should be a better way to determine the value since structure packing is implementation defined.

The test in main function results non-determinate order of jobs echoing its id. A mechanism to wait all jobs to complete should be added later instead of using sleep. `thread_pool_destroyed()` is not functional yet.

To explain `Exchange`, `Test and set`, `Fetch and ...` and `Compare and swap` in section 5.1~5.4 using this example, the following issues should be resolved:
* `Exchange` is not in use.
* `Test and set` on the `initialized` flag is useless. Testing the flag on the first thread pool initialization retrieve non-determinate value. Currently re-initialization is still possible. If preventing re-initialization is not crucial in this example, we need to find another way to use test and set.
* Using `Fetch and add` to change thread pool state seems too intentional.
* Pick a better job type, such as BBP PI approximation.
* Finish implementation mentioned in the above paragraph.

Should we break the example into pieces to explain individually? Or list the code first?

idoleat committed:
Add new example for chapter 5

A new directory /examples is added for example codes. .clang-format is
copied from sysprog21/lkmpg project.

This commit provides a simplified implementation of thread pool. After
initializing the thread pool with thread count, jobs can be added. The
job queue is a SPMC ring buffer. To keep the implementation minimal,
the producer is not protecting, resulting the thread pool can not run
automatically when jobs are added. Or the worker may try to get the job
before it is fully enqueued.

Padding is added in thread_pool_t to avoid false sharing. The number
"40" is the sum of size of struct members, including alignment, before
the first padding. There should be a better way to determine the value
since structure packing is implementation defined.

The test in main function results nondeterminate order of jobs echoing
its id. A mechanism to wait all jobs to complete should be added later
instead of using sleep. `thread_pool_destroyed()` is not functional yet.

jserv commented on examples/ch5_example.c:
```diff
@@ -0,0 +1,151 @@
+
+typedef struct job {
+    void *args;
+    struct job *next;
```
Declare as `struct job *prev, *next;`

jserv commented on examples/ch5_example.c:
```diff
@@ -0,0 +1,151 @@
+
+    while (1) {
+        uintptr_t job;
+        if (atomic_load(&thrd_pool->state) == running) {
```
Rewrite as following:
```c
if (atomic_load(&thrd_pool->state) == cancelled)
    return EXIT_SUCCESS;
if (atomic_load(&thrd_pool->state) == running) {
    ....
} else {
    continue;
}
```

idoleat committed:
Refactor worker thread conditional code path

Codes executed under running condition are placed in the same scope.

idoleat committed:
Combine identifier declarations in one line

Both next and prev in job_t are struct job, thus residing them in the
same line.

jserv commented on examples/ch5_example.c:
```diff
@@ -36,32 +36,30 @@ int worker(void *args)
-        if (atomic_load(&thrd_pool->state) == running) {
+        if (atomic_load(&thrd_pool->state) == cancelled) {
+            return EXIT_SUCCESS;
+        } else if (atomic_load(&thrd_pool->state) == running) {
```
Here `else if` can shortened as a `if` statement.

jserv commented on examples/ch5_example.c:
```diff
@@ -0,0 +1,148 @@
+                    // as long as producer is protected
+                }
+            } else {
+                printf("Hello from job %d\n", *(int *)(*(job_t **)job)->args);
```
Can you avoid using `*(int *)(*(job_t **)job)` by typedef?

idoleat replied:
I think using macro could be better?
```c
#define CAST_JOB(job, type) ({ \
    *(typeof(type) *)(*(job_t **)(job))->args; \
})
```
so args in `printf` could be
```c
printf("Hello from job %d\n", CAST_JOB(job, int));
```

idoleat replied:
Also the above `(uintptr_t)(&(*(job_t **)job)->prev)` could be simplified to
```c
atomic_compare_exchange_strong(&thrd_pool->tail, &job, PREV_JOB(job))
```
by using
```c
#define PREV_JOB(job) ({ (uintptr_t)(&(*(job_t **)(job))->prev); })
```

jserv commented on examples/ch5_example.c:
```diff
@@ -0,0 +1,148 @@
+    atomic_flag_test_and_set(&thrd_pool->initialezed); // It's useless anyway
+
+    // TODO: size should be a positive integer
+    // malloc with zero size is non-portable
```
Simply place an `assert` to avoid the case size equals zero.

idoleat committed:
Add assert to avoid the case size equals zero

An assert is added before malloc. Also type of size to changed to
size_t.

idoleat committed:
Shorten else if as an if

idoleat committed:
Replace long expressions with macros

Two macros `CAST_JOB(job, type)` and `PREV_JOB(job)` are added to
simplify long expressoins to improve readability.

idoleat commented:
The job queue exhibits incorrect behavior and is currently under rework. To reproduce the incorrect behavior, add new jobs after finishing the current ones. A segmentation fault occurs simply because `thrd_pool->tail` is an independent variable, not a pointer to the real tail. It still holds the old address when new jobs are added. I overthought the situation, assuming that `prev` in `job_t` would be constantly changed, and thus made it a distinct struct member, even with padding.

idoleat committed:
Fix wrong behavior of job queue

A new struct `idle_job` is added. There are several ways to have atomic
`thrd_pool->head->prev` (the original tail):

(1) _Atomic used as either specifier or qualifier in C11 acts on object, not
region. So we can not have only the idle `job_t` have atomic `prev`. All
job_t would have atomic `prev`. However _Atomic is only allowed to act
on complete type, meaning that `_Atomic(struct job *)` and `_Atomic(void
*)` are not allowed in the declaration of `job_t`. `atomic_uintptr` has already
shown enough casting chaos in the previous commit.

(2) Embed `job_t` in a new struct `idle_job` along side `_Atomic(job_t *) prev;`.
In worker function, the last job is accessed through
`thrd_pool->head->prev`, the same as (1). The only difference is how the
idle job is initialized and how the first job is added. Padding could
also be added around `prev` to avoid false sharing.

Test in main demonstrates a series of jobs are added after fininshing
existing ones. `thread_pool_destroy` is implemented to cancel and free the
pool. Additional sleep could be added before detroy to observ the second
series of jobs.

Notice that freeing memory of job in worker directly after using it may
cause dangling pointer in other threads. Safe memory reclaimation should
be introduced to avoid this completely. Or use memory pool for jobs.

idoleat committed:
Add use case of atomic exhange, TAS, fetch AND

Atomic flag is used for checking if the given thread pool has been
initialized. The flag is initialized when thread pool struct declared and
reseted to false when thread pool destroyed.

Atomic exchange obtains previous state when destroying thread pool and gives
warnning message if the state were running.

Atomic fetch and AND with zero demonstrate a way to set state to idle.

idoleat commented:
> the following issues should be resolved:
> * `Exchange` is not in use.
> * `Test and set` on the `initialized` flag is useless. Testing the flag on the first thread pool initialization retrieve non-determinate value. Currently re-initialization is still possible. If preventing re-initialization is not crucial in this example, we need to find another way to use test and set.
> * Using `Fetch and add` to change thread pool state seems too intentional.

Solved.

Operations on atomic types, including flag type, covered in C11 standard 7.17.7 and 7.17.8 are all used in the example code. Next I will start revising section 5 based on this example. The undesired result caused by not using atomic operations will be provided as well. Also clarifying more on why we need read-modify-write as a atomic step.

jserv commented on examples/ch5_example.c:
```diff
@@ -0,0 +1,169 @@
+int main()
+{
+    thread_pool_t thrd_pool = { .initialezed = ATOMIC_FLAG_INIT };
+    int thread_count = 8;
```
Both `job_count` and `thread_count` were meant to be constant, and you can simply specify them as macros. That is, `N_JOBS` and `N_THREADS`.

idoleat committed:
Define job count and thread count as constant

Both job_count and thread_count were meant to be constant in the given
test scenario. Thus they were specified as macros instead of variables.

idoleat committed:
Revise section 5 with new example

Clarification on read-modify-write is first added at the beginning of
the section. To make the discussion based on atomic load/store, more
information is supplemented at the end of section 2.

Example code is included by using minted package. Each subsection is
revised according to the atomic library usage in the example.

At the end of sectoin 5, a new subsection "further improvements" is
added to discuss topics on leveraging other memory order, false sharing
and safe memory reclamation. First two topics are forward referenced to
according chapters. The last one is not covered in this book so it has no
reference.

idoleat committed:
Rename ch5_example.c to rmw_example.c

rmw_cample.c is committed in the last commit.

idoleat committed:
Remove unneeded comment

idoleat commented on concurrency-primer.tex:
```diff
@@ -321,6 +321,9 @@ \section{Enforcing law and order}
 This model, defined by Leslie Lamport in 1979,
 is called \introduce{sequential consistency}.
 
+Notice that using atomic variables as an lvalue expression, such as \monobox{v\_ready = true} and \monobox{while(!v\_ready)}, is a convenient alternative to explicitly using \monobox{atomic\_load} or \monobox{atomic\_store}. 
```
Code generation on atomic load and store is a little bit tricky. As explained in [LLVM's document](https://llvm.org/docs/Atomics.html#id16), sometimes they are just normal load/store, without atomic instructions, and their code generation can vary across different architectures. Should we explain more on this? Or readers may be confused on examples in section 7 and 10.1

jserv replied:
Can you summarize the above in the footnote?

idoleat replied:
Resolved in 4dc5699.

idoleat commented on concurrency-primer.tex:
```diff
@@ -321,6 +321,9 @@ \section{Enforcing law and order}
 is called \introduce{sequential consistency}.
 
+Notice that using atomic variables as an lvalue expression, such as \monobox{v\_ready = true} and \monobox{while(!v\_ready)}, is a convenient alternative to explicitly using \monobox{atomic\_load} or \monobox{atomic\_store}. 
+Lvalue-to-rvalue conversion (which models a memory read from an atomic location to a CPU register) strips atomicity along with other qualifiers. 
```
> This is from C11 6.7.2.4 and 6.7.3. Should I add reference as a footnote?

No, simply add in the text body.

idoleat replied:
Resolved in 4dc5699.

idoleat commented on concurrency-primer.tex:
```diff
@@ -443,52 +464,54 @@ \subsection{Compare and swap}
-template <typename T>
-bool atomic<T>::compare_exchange_strong(
-    T& expected, T desired)
+\begin{ccode}
```
This is borrowed from C11 7.17.4. Should I add reference as a footnote?

jserv replied:
> This is borrowed from C11 7.17.4. Should I add reference as a footnote?

Add the proper statement.

idoleat replied:
Resolved in 4dc5699.

idoleat commented on concurrency-primer.tex:
```diff
@@ -443,52 +464,54 @@ \subsection{Compare and swap}
-};
+Once threads are created in the thread pool as workers, they will continuously search for jobs to do.
+Jobs are taken from the tail of job queue.
+To claim a job without it being taken by another worker halfway through, we need to atomically change the pointer to the last job. Otherwise the last job is under races.
```
Should provide an example result of races with `-fsanitize=thread -g` enabled.

idoleat replied:
Thread sanitizer currently does not support C11 thread `[1]` `[2]` `[3]`*. Using functions in `threads.h` with `-fsanitize=thread` results in segmentation fault in thread sanitizer runtime. I will provide a minimum modification in git diff format in this subsection to run the example with Tsan using `pthread.h` instead.

`[1]`: https://clang.llvm.org/docs/ThreadSanitizer.html
`[2]`: https://github.com/google/sanitizers/wiki/ThreadSanitizerCppManual#faq
`[3]`: https://github.com/intel/libipt/commit/f24d71340f70dcdc2f738333bc6c09cec6509a81
*: Both documents [1] and [2] don't clearly specify that Tsan doesn't support C11 thread yet. By most-recent commit confirming this [3] and C++11 thread still requiring rebuilding libc++ with Tsan, I believe C11 is not supported in Tsan yet.

idoleat committed:
Add race conditoin example

A diff file is added to patch original example to the one that can cause
races. Substituting `threads.h` with `pthread.h` is also included in
diff because sanitizer hasn't support C11 thread yet. How sanitizer
works and how to use it are added as well. Then explainations on
warning messages from Tsan are followed.

The part mentioning safe memory reclaimation is moved to this subsection
because warning messages from Tsan mentioned it.

A missing reference to spinlock (originally as one of the rmw examples)
in section 9 was added back as a new code block.

idoleat commented on concurrency-primer.tex:
```diff
@@ -706,7 +753,17 @@ \section{Do we always need sequentially consistent operations?}
 What if we could avoid some of this slowdown?
-Consider a simple case like the spinlock from \secref{spinlock}.
+Consider a simple case like spinlock.
+\begin{ccode}
```
`\secref{spinlock}` is originally one of the examples in section 5 and replaced by new example. Commit 2caa7c0 solved the missing reference by unrolling the example.

This is not a good example and is not simple. Spinning on RMW operation can cause cache bouncing. Test-and-test-and-set pattern is one way to solve this problem. Functionally, this is correct, but it may mislead readers by oversimplifying spin lock mechanics and ignoring the details behind it.

Spin lock topic should be moved to cache effect section. Here we simply use another example instead. Not sure if scheduling related topic, such as sched_yeild or spinning in user space, should be covered though. If so, a section dedicated to scheduling may be needed.

idoleat replied:
> Here we simply use another example instead

In 926f645,  spinlock example is replaced with new example provided in section 5.

idoleat replied:
Resolved as a6c68a9 restored the original content.

idoleat committed:
Refine the statement on atomicity

The original statement is only true on the successful operation. To 
other failed operations, it is the successful one that finished before 
failed ones.

The new statement does not totally cover the charateristic of atomic 
operations though. It is the generated cmpxchg or LL/SC loop that make 
the operation keeps trying and eventually finish. But considering that 
the purpose of this paragraph is to plot a big picture of order and 
atomicity, more details on atomic operations should be covered in 
section "Atomic as building blocks". More refences to compiler and CPU 
menufacturer documents should be taken in consideration then.

idoleat committed:
Update descriptions referencing to section 5

The intro of section 10 originally references back to spinlock in 
section 5. It is now replace by using the new example.

Section 10.2 onriginally  references back to UI thread in section 5. It 
is now removed as new example presents down below to explain relaxed 
memory model. New example is used here as well is because it is exactly 
what original example was talking about.

idoleat commented on concurrency-primer.tex:
```diff
@@ -853,51 +900,41 @@ \subsection{Relaxed}
-
-Relaxed operations are also beneficial for managing flags shared between threads.
-For example, a thread might continuously run until it receives a signal to exit:
+Relaxed operations are beneficial for managing flags shared between threads.
```
Should we mention the discussion around relaxed atomics?
like this one: https://lukegeeson.com/blog/2023-10-17-A-Proposal-For-Relaxed-Atomics/

idoleat committed:
Refine rmw example and remove sleep()

Static linkage is added for better practice. A new inline function 
wait_until is added to serve the need of waiting thread pool until give 
state, thus removing sleep() and corresponding header.

The weak version of compare and swap is used instead due to
1. There is really no other thing in the same cache line to cause 
   spurious fail.
2. The retry cost is considered lower than nested loop

weihsinyeh commented:
To demonstrate the application of Test and Set in rmw_example.c:

As stated in the Concurrency Primer (§ 5.2) regarding Test and Set: "We could use Test and Set to build a simple spinlock."

In rmw_example.c, Compare and Swap is utilized to avoid race conditions. However, the Test and Set operation using atomic_flag_test_and_set() is only employed for initialization, which underutilizes its locking properties to avoid race condition.

To illustrate the distinct applications of Test and Set versus Compare and Swap:
Test and Set : If the Test operation fails, it will spin. 
Compare and Swap : When a Compare operation fails, it will continue to execute.

However, to discuss this in rmw_example.c, another shared resource must be introduced. Currently, the shared resource is a thrd_pool.

Consider a scenario where multiple workers all need to access the same shared variable to perform operations. In this case, mutex can be utilized to provide the functionality similar to a Test and Set flag.

In rmw_example.c, "Exchange, Test and Set, Fetch and ..., and Compare and Swap" performs all four operations using atomic operations from stdatomic.h. However, as mentioned in Chapter 1: "System programmers are familiar with tools such as mutexes, semaphores, and condition variables. Nevertheless, a question remains: How do these tools function, and how can we write concurrent code in their absence?"

Therefore, it is not mandatory to use atomic_flag_test_and_set() for Test and Set. Mutex can also achieve locking and unlocking effects.

weihsinyeh commented:
Another idea for Chapter 5 is to retain the original structure, keeping the descriptions of concepts in sections 5.1, 5.2, 5.3, and 5.4. However, the original examples are not intuitive because all of these sections are actually related to RMW operations. It's important for readers to understand that they only differ in techniques, aimed at being applicable to different scenarios. At the same time, specifically explain where exactly the need for atomic operations arises.

Implementing RMW does not necessarily require tools like those found in stdatomic.h.

In a subsequent new subsection 5.5, further explanation will detail the atomic functions available to implement these four concepts. In rmw_example.c, demonstrate their application in specific scenarios.

weihsinyeh committed:
Add BBP formula to approximate PI

1. Use Bailey–Borwein–Plouffe formula to approximate PI
-  Reference : https://github.com/sysprog21/concurrent-programs/blob/master/tpool/tpool.c

2. Add Add PRECISION constant with value 100

weihsinyeh committed:
Worker do the job to execute BBP

1. Add the tpool_future variable
- tpool_future to pass the result to the main thread.
- The mutex lock and the condition variable to ensure concurrency.

2. The main thread sequentially accumulate results from BBP that
calcuate by every worker.
- Wait using `tpool_future_get()` until the condition variable is
broadcast to confirm that the result has been marked as
__FUTURE_FINISHED.

3. Change `thread_pool` to `tpool` to improve readability.

4. Add the Makefile.

weihsinyeh committed:
Change the implementation of mutex to test and set

1. Directly show the scenarios using Test and Set and
its atomic operations.
- Use `atomic_flag_test_and_set()` and `atomic_flag_clear()` to
implement the original mutex lock and unlock mechanism.
- Replace the original condition variable wait mechanism with
`atomic_flag_test_and_set()` combined with a `while` loop.

2. Avoid deadlock in `tpool_future_get()`.
- The main thread must first wait for the worker to complete the "BBP
formula" job.
- Subsequently, it should wait for the worker to unlock.
- These two operations must occur in this order to avoid deadlock.
Swapping them will lead to deadlock.

weihsinyeh committed:
Check if job is finished by the result of future

1. Check if `future->result` is NULL.
- If `future->result` is NULL, the job is still in progress.
- If `future->result` is not NULL, the job has been completed by the
worker.

weihsinyeh committed:
Avoid the memory leak in the add_job function

1. When allocating memory for future, if the allocation fails, do not
simply return NULL. Instead, release the memory allocated for job
beforehand to avoid memory leaks.

jserv reviewed (changes requested):
Rebase the latest `main` branch for reviewing.

weihsinyeh committed:
Simplify the operation of test and set

1. When creating the future, set the future's flag, which is akin to
assigning the job. Afterward, transfer the ownership to the worker.
Once the worker completes the job, clear the flag and return the
ownership, which is akin to submitting a job. Then, the main thread
can regain ownership. By doing this, the main thread can wait directly
for the result through test and set without checking if the result is
NULL. This avoids the situation where the flag could be set to true by
the main thread before the worker starts the job. Additionally, the
worker does not need to check with test and set before performing the
job.

2. Drop the `atomic_flag_clear` in `tpool_future_wait` function and then
directly free the pointer of future and its result in
`tpool_future_destroy` function.

3. Rename the variable 'lock' in the future structure to 'flag'. Rename
the function name `tpool_future_get` to `tpool_future_wait`.

Co-authored-by: Chih-Wei Chien <idoleat@taiker.tw>
Signed-off-by: Wei-Hsin Yeh <weihsinyeh168@gmail.com>

weihsinyeh committed:
Avoid memory leak in the bbp function

1. When allocating memory for the product, if the allocation fails,
it returns NULL.

Co-authored-by: Chih-Wei Chien <idoleat@taiker.tw>
Signed-off-by: Wei-Hsin Yeh <weihsinyeh168@gmail.com>

jserv commented on concurrency-primer.tex:
```diff
@@ -452,52 +473,78 @@ \subsection{Compare and swap}
+When races occur during execution, warning messages are printed.
+To enable this, add \monobox{-fsanitize=thread -g} to compiler flags.
+The following diff patch removes the atomicity of claiming a job and uses pthread instead of \clang{}11 thread, because thread sanitizer currently hasn't support \clang{}11 thread yet. 
+Save diff as \monobox{racer.diff} and patch the example code by \monobox{\$ patch rmw\_example.c race.diff}.
```
Don't do that. This is a book, and we should emphasize the readability.

idoleat replied:
Got it. Would it be a better way to use conditional macros in the example code? like this:
```c
typedef struct tpool {
    atomic_flag initialezed;
    int size;
    atomic_int state;
#ifdef PTHREAD
    pthread_t *pool;
#else
    thrd_t *pool;
#endif
    idle_job_t *head;
} tpool_t;
```
then use a short a description guiding the reader to edit 3 lines of code to create race condition.

jserv replied:
Let's stick to C11 Threads rather than POSIX Thread.

idoleat replied:
Resolved in dcb8f9a.

weihsinyeh committed:
Integrate RMW with concepts from previous sections

1. Use 2 figures to connect concepts from the first 3 sections.
- Figure atomic_rmw illustrates that atomic operations consist of not
only a single operation but a group of operations that need to perform
atomically.

- Figure rmw_communicate shows how this atomic group of operations can
be used on shared resource for communication.

2. Discuss how to ensure the operations of accessing the shared resource
for communication between concurrent threads are correct:
- Use Test and Set and Compare and Swap as examples to illustrate how
this can be achieved.

3. Compare the usage scenarios of Exchange and Fetch and ...

4. Introduce the concept that we can utilize atomic operations to
ensure that a group of operations can perform atomically.

idoleat committed:
Remove the thread santizer usage and diff file

Introducing thread sanitizer here may be an unexpected pop up for the
readers that are new to concurrency. Here we focus on rmw atomic
operation instead, thus related content and diff file are removed.
The proper place for this topic could be a dedicated section for
"testing, debugging and verifing concurrent programs"

This aligns the decision sticking to C11 thread as well.


Co-authored-by: Wei-Hsin Yeh <weihsinyeh168@gmail.com>

idoleat committed:
Restore previously removed spinlock example

Since spinlock is added back in section 5.2, the original content is 
restored. Same as the rmw example, the goal is to provide easy to 
understand example first and improve it later on.

jserv commented on concurrency-primer.tex:
```diff
@@ -380,39 +383,59 @@ \section{Arbitrarily-sized ``atomic'' types}
+
+\includegraphics[keepaspectratio, width=0.6\linewidth]{images/atomic_rmw}
+\captionof{figure}{Exchange, Test and Set, Fetch and…, Compare and Swap can all be transformed into atomic RMW operations, ensuring that operations like t1 \to t2 \to t3 will become an atomic step.}
+\label{fig:atomic_rmw}
```
Use `fig:atomic-rmw` for consistency.

idoleat replied:
Resolved in 11baa72.

jserv commented on concurrency-primer.tex:
```diff
@@ -431,73 +454,124 @@ \subsection{Test and set}
+or bitwise \textsc{AND}, \textsc{OR}, \textsc{XOR}) and return its previous value, 
+all as part of a single atomic operation. 
+Compare with \introduce{Exchange} \secref{exchange}, when programmers only need to make simple modification to the shared variable, 
+they can use \introduce{Fetch and…}.
```
It is weird to have `Fetch and…`

weihsinyeh replied:
Could you please explain more?

idoleat committed:
Supplement statements with references to C11 and LLVM docs

References to C11 standard were added when explaining properties of
atomic type and operations. More information of codegen on atomic
operations is added as a footnote with a link to LLVM's document as an
example.

jserv commented on examples/rmw_example.c:
```diff
@@ -0,0 +1,241 @@
+
+    for (size_t i = 0; i < size; i++) {
+        thrd_create(thrd_pool->pool + i, worker, thrd_pool);
+        //TODO: error handling
```
We may ignore some error handling for the sake of providing minimal examples.

weihsinyeh committed:
Add atomic instruction and simplify rmw_example.c

Add the description of atomic instruction to let readers know there is a
difference between using fetch and..., which is only a programming tool,
and its actual execution as an atomic operation that depends on the
compiler.

Simplify the rmw_example code to provide more flexible examples.

 - Initially, all worker threads will be initialized. The main thread
will ask all workers to start running. If there is no job or the job is
completed, the worker will become idle. Next, the main thread will
continue to add more jobs and ask the worker to start running again.
Meanwhile, the main thread will also wait for the results of the work.

 - Use the struct `tpool_future` to record all the information required
for the job.

Co-authored-by: Chih-Wei Chien <idoleat@taiker.tw>

jserv commented:
Thank @idoleat for contributing!
